### PR TITLE
ipam and idm for multi host

### PR DESCRIPTION
**Pull request by aboch**
Provides a new ip address management package for multi host networking

### CODE REVIEW

File: bitseq/sequence.go [hunk 1]
```diff
@@ -0,0 +1,450 @@
...
+	Head       *Sequence
+	App        string
+	ID         string
+	dbIndex    uint64
+	store      datastore.DataStore
+	sync.Mutex
+}
+
+// NewHandle returns a thread-safe instance of the bitmask handler
+func NewHandle(app string, ds datastore.DataStore, id string, numElements uint32) *Handle {
```
Comment (mrjana) [inline bitseq/sequence.go]: This api needs to handle errors most typically emanating from data store, like if we fail to connect to the data store.

File: bitseq/sequence.go [hunk 2]
```diff
@@ -0,0 +1,450 @@
...
+	blockBytes    = blockLen / 8
+	blockMAX      = 1<<blockLen - 1
+	blockFirstBit = 1 << (blockLen - 1)
+)
+
+// Handle contains the sequece representing the bitmask and its identifier
+type Handle struct {
+	bits       uint32
+	unselected uint32
+	Head       *Sequence
```
Comment (mrjana) [inline bitseq/sequence.go]: make this unexported so that nobody can manipulate the sequence directly

File: ipam/allocator.go [hunk 3]
```diff
@@ -0,0 +1,610 @@
...
+	// Datastore
+	store   datastore.DataStore
+	App     string
+	ID      string
+	dbIndex uint64
+	sync.Mutex
+}
+
+// NewAllocator returns an instance of libnetwork ipam
+func NewAllocator(ds datastore.DataStore) *Allocator {
```
Comment (mrjana) [inline ipam/allocator.go]: Again need to check for error and return error from data store

File: ipam/allocator.go [hunk 4]
```diff
@@ -0,0 +1,610 @@
...
+
+	// Now retrieve the correspondent address masks from the store
+	a.Lock()
+	for k, v := range a.subnets {
+		subnetList, err := getInternalSubnets(v.Subnet, a.internalHostSize)
+		if err != nil {
+			log.Warnf("Failed to load address bitmask for configured subnet %s because of %s", v.Subnet.String(), err.Error())
+			continue
+		}
+		a.Unlock()
```
Comment (mrjana) [inline ipam/allocator.go]: Please refactor this code to no unlock it in the middle the map traversal

File: ipam/allocator.go [hunk 5]
```diff
@@ -0,0 +1,610 @@
...
+	}
+
+	// Register for status changes
+	a.watchForChanges()
+
+	// Get the initial subnet configs status from the ds if present.
+	kvPair, err := a.store.KVStore().Get(datastore.Key(a.Key()...))
+	if err != nil {
+		log.Warnf("Failed to retrieve the ipam subnet configs from datastore")
+		return a
```
Comment (mrjana) [inline ipam/allocator.go]: if there is a failure you should be returning a nil allocator

File: ipam/allocator.go [hunk 6]
```diff
@@ -0,0 +1,610 @@
...
+	a.subnets[key] = subnetInfo
+	a.Unlock()
+
+	// Sync to datastore
+	err = a.writeToStore()
+	if err != nil {
+		// Update to latest in any error case
+		a.readFromStore()
+		if _, ok := err.(types.RetryError); !ok {
+			return types.InternalErrorf("subnet configuration failed because of %s", err.Error())
```
Comment (mrjana) [inline ipam/allocator.go]: You need to remove `a.subnets[key]` if read fails

File: ipam/allocator.go [hunk 7]
```diff
@@ -0,0 +1,610 @@
...
+	delete(a.subnets, subKey)
+	a.Unlock()
+
+	// Sync to datastore
+	err := a.writeToStore()
+	if err != nil {
+		// Update to latest in any error case
+		a.readFromStore()
+		if _, ok := err.(types.RetryError); !ok {
+			return types.InternalErrorf("subnet removal failed because of %s", err.Error())
```
Comment (mrjana) [inline ipam/allocator.go]: Same comment as above for `AddSubnet` where the read fails you need to rollback to the previous state

File: ipam/allocator.go [hunk 8]
```diff
@@ -0,0 +1,610 @@
...
+		if _, ok := err.(types.RetryError); !ok {
+			return nil, fmt.Errorf("internal failure while reserving the address: %s", err.Error())
+		}
+		// bitmask view must have changed. Selected address may or may no longer be available
+		if prefAddress != nil {
+			if _, _, err = bitmask.CheckIfAvailable(ordinal); err == nil {
+				//still available
+				goto pushsame
+			}
+			goto again
```
Comment (mrjana) [inline ipam/allocator.go]: Can we please refactor this code to not use goto this way of jumping back? Instead a for loop with `continue` and `break

File: bitseq/sequence.go (hunk 1 above)
Comment (aboch) [inline bitseq/sequence.go]: Thanks, will change it.

File: bitseq/sequence.go (hunk 2 above)
Comment (aboch) [inline bitseq/sequence.go]: Good idea.

File: ipam/allocator.go (hunk 3 above)
Comment (aboch) [inline ipam/allocator.go]: ok

File: ipam/allocator.go (hunk 4 above)
Comment (aboch) [inline ipam/allocator.go]: I will

File: ipam/allocator.go (hunk 5 above)
Comment (aboch) [inline ipam/allocator.go]: Ok

File: ipam/allocator.go (hunk 6 above)
Comment (aboch) [inline ipam/allocator.go]: Good catch, I ignored read error

File: ipam/allocator.go (hunk 8 above)
Comment (aboch) [inline ipam/allocator.go]: Ok

File: idm/idm.go [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+import (
+	"fmt"
+
+	"github.com/docker/libnetwork/bitseq"
+	"github.com/docker/libnetwork/datastore"
+)
+
+// Idm manages the reservation/release of numerical ids from a contiguos set
+type Idm struct {
+	start  int
```
Comment (mrjana) [inline idm/idm.go]: Can you please change the idm start, end and the return values of `GetID` and other apis to uint32?

Comment (aboch) [inline idm/idm.go]: Good idea. Since it's for generic ids, let's make it uint64

File: idm/idm.go [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+		return 0, fmt.Errorf("no available ids")
+	}
+	id := i.start + bitPos + bytePos*8
+
+	// for sets which length is non multiple of 32 this check is needed
+	if i.end < id {
+		return 0, fmt.Errorf("no available ids")
+	}
+
+	i.handle.PushReservation(bytePos, bitPos, false)
```
Comment (mrjana) [inline idm/idm.go]: This needs to handle retryerror

File: idm/idm.go [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+		return fmt.Errorf("ID set is not initialized")
+	}
+
+	if id < i.start || id > i.end {
+		return fmt.Errorf("Requested id does not belong to the set")
+	}
+
+	if bytePos, bitPos, err := i.handle.CheckIfAvailable(id - i.start); err == nil {
+		i.handle.PushReservation(bytePos, bitPos, false)
+		return nil
```
Comment (mrjana) [inline idm/idm.go]: This needs to handle retry error

**Comment by mrjana** [comment]
LGTM

**Pull request merged**